### PR TITLE
Query support

### LINKED ISSUE #25: Add query support
Input: text field to input JSON
Output: query response
API: `CosmWasmClient.queryContractSmart`

### PR DESCRIPTION
**Pull request by abefernan**
Closes #25

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,6 +32,7 @@
     ]
   },
   "dependencies": {
+    "@cosmicdapp/logic": "^0.4.5",
```
Comment (webmaster128) [inline package.json:35]: This is unused, right? I think you can just remove the "Adds new dependencies" commit from this PR

File: src/pages/contract/QueryContract.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            className="ml-3 flex-grow-1"
+            value={queryInput}
+            onChange={(event) => setQueryInput(event.target.value)}
+          />
+        </li>
+        {queryObject?.result ? (
+          <>
+            <li className="list-group-item">
+              <span title="The contract formatted input">Formatted query:</span>
+              <pre className="mt-2 mb-3">{JSON.stringify(queryObject.result, null, "  ")}</pre>
```
Comment (webmaster128) [inline src/pages/contract/QueryContract.tsx:74]:
```suggestion
              <pre className="mb-0">{JSON.stringify(queryObject.result, null, "  ")}</pre>
```

No need for extra spacing I think. list-group-item has a padding already

File: src/pages/contract/QueryContract.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            <li className="list-group-item">
+              <button className="btn btn-primary" onClick={runQuery} disabled={!!error}>
+                Run query
+              </button>
+            </li>
+          </>
+        ) : null}
+        {queryResponse?.result ? (
+          <li className="list-group-item">
+            <span title="The contract formatted input">Response:</span>
```
Comment (webmaster128) [inline src/pages/contract/QueryContract.tsx:85]:
```suggestion
            <span title="The query response">Response:</span>
```

**Review by webmaster128 — COMMENTED** [review]
Looking good.

I'm not sure if we really need the formatted input, but I don't have a better idea right now. Ideally we find a multi-line JSON input component with highlighting an in-line errors. Then one field would be enought.

**Comment by abefernan** [comment]
> Ideally we find a multi-line JSON input component with highlighting an in-line errors. Then one field would be enought.

I found this, should I integrate it in this PR?: https://www.npmjs.com/package/react-json-editor-ajrm

**Comment by webmaster128** [comment]
> I found this, should I integrate it in this PR?: https://www.npmjs.com/package/react-json-editor-ajrm

Cool! Yeah, let's give it a try

**Review by webmaster128 — APPROVED** [review]

**Review by webmaster128 — COMMENTED** [review]
Waiting for the new input component before merge

File: src/pages/contract/QueryContract.tsx [hunk 4]
```diff
@@ -61,25 +50,33 @@ export function QueryContract({ contractAddress }: Props): JSX.Element {
...
-          <input
-            className="ml-3 flex-grow-1"
-            value={queryInput}
-            onChange={(event) => setQueryInput(event.target.value)}
+        </li>
+        <li className="list-group-item d-flex align-items-baseline">
+          <JSONInput
+            width="100%"
+            height="200px"
+            placeholder={{ field: "value" }}
```
Comment (webmaster128) [inline src/pages/contract/QueryContract.tsx:58]: Can we use this as a default?

```json
{
  "get_balance": {
    "address": "cosmos1zk4hr47hlch274x28j32dgnhuyewqjrwxn4mvm"
  }
}
```

File: src/pages/contract/QueryContract.tsx [hunk 5]
```diff
@@ -61,25 +50,33 @@ export function QueryContract({ contractAddress }: Props): JSX.Element {
       <ul className="list-group list-group-flush">
         <li className="list-group-item d-flex align-items-baseline">
           <span title="The contract query input">Query contract:</span>
-          <input
-            className="ml-3 flex-grow-1"
-            value={queryInput}
-            onChange={(event) => setQueryInput(event.target.value)}
+        </li>
+        <li className="list-group-item d-flex align-items-baseline">
+          <JSONInput
```
Comment (webmaster128) [inline src/pages/contract/QueryContract.tsx:55]: Can we configure this component to only accept JSON, not all JavaScript?

Comment (webmaster128) [reply src/pages/contract/QueryContract.tsx:55]: Answer seems to be no. See e.g. https://github.com/AndrewRedican/react-json-editor-ajrm/issues/120

**Pull request merged**